### PR TITLE
feat: make previews great again

### LINKED ISSUE #1457: [Feature UI]: Dynamic OG previews for shared Pubky links
# Context

Comes from [Northstar Roadmap #38](https://northstar-eta.vercel.app/?task=38) | Due: Mar 20 | Absorbs [Northstar #264](https://northstar-eta.vercel.app/?task=264)

## Background

When you share a Pubky link on Twitter, Slack, Discord, or iMessage, every single link shows the same generic "Unlock the web" keyhole image (`/public/preview.webp`). Doesn't matter if it's a post, a profile, the hot page. Same image, always.

Posts do have dynamic title/description via `generateMetadata()` in `src/app/post/[userId]/[postId]/page.tsx` (added Jan 29, commit `7a0050a3`). So Slack will show "John on Pubky" with the post text, which is something. But the image is still the generic one, and profiles have no metadata at all.

The roadmap calls this "Everything shareable" and it's tagged as both Design and Engineering. Aldert needs to define what the OG cards look like, then someone builds the image generation.

## Current state

| Page | Title/Description | og:image | Notes |
|------|-------------------|----------|-------|
| `/post/[userId]/[postId]` | Dynamic (username, post snippet) | Static `preview.webp` | `generateMetadata()` exists |
| `/profile/[pubky]` | Falls back to default | Static `preview.webp` | No `generateMetadata()` at all |
| `/hot`, `/search`, `/` | Default "Pubky App" | Static `preview.webp` | |

The `Metadata()` helper in `src/components/molecules/Metadata/Metadata.tsx` already accepts an `image` param, it just defaults to `Config.DEFAULT_PREVIEW_IMAGE`. So the plumbing is there, we just need something to generate the images.

No `@vercel/og` or Satori in `package.json`. No `/api/og-image` route. No `opengraph-image.tsx` convention files.

### 1. Design: OG card templates (Aldert)

Before anyone writes code, we need designs for what these cards look like. At minimum:
- **Post card**: author avatar, name, post text snippet, Pubky branding. Dark theme matching the app.
- **Profile card**: avatar, name, bio, follower count, Pubky branding.
- **Image post variant**: use the actual post image instead of the generic template.

These are 1200x630 images. They need to look good at thumbnail size on Twitter/Slack/Discord.

### 2. Dynamic OG image endpoint (Dev)

Create `/api/og-image` route using `@vercel/og` (Satori). It takes query params like `type=post&userId=X&postId=Y`, fetches the data server-side, renders the card to a PNG.

Files:
- New: `src/app/api/og-image/route.tsx`
- Install: `@vercel/og` (or use Next.js built-in `ImageResponse` from `next/og`)

### 3. Wire up post pages (Dev)

`src/app/post/[userId]/[postId]/page.tsx` already has `generateMetadata()`. Just need to change the `image` param from the static webp to the dynamic endpoint. Something like:

```
/api/og-image?type=post&userId=${userId}&postId=${postId}
```

For image posts, pass the image URL so the card can use it directly.

### 4. Add profile metadata (Dev)

`src/app/profile/[pubky]/page.tsx` has zero metadata. Add `generateMetadata()` that fetches user details and points og:image at:

```
/api/og-image?type=profile&userId=${userId}
```

### 5. Validate on platforms

Test the previews actually render correctly on Twitter card validator, Slack, Discord, iMessage, WhatsApp. Each platform has quirks with caching and size limits.

## Acceptance criteria

- [ ] Design: OG card mockups for post and profile (Aldert sign-off)
- [ ] `@vercel/og` or `next/og` added, `/api/og-image` route working
- [ ] Post links show dynamic card with author, avatar, content snippet
- [ ] Image posts use the actual image in the OG card
- [ ] Profile links show name, bio, avatar, follower count
- [ ] All generated images are 1200x630, cached at CDN level (1h+)
- [ ] Tested on Twitter, Slack, Discord, iMessage

## Files

- Post page metadata: `src/app/post/[userId]/[postId]/page.tsx`
- Profile page (no metadata yet): `src/app/profile/[pubky]/page.tsx`
- Metadata helper: `src/components/molecules/Metadata/Metadata.tsx`
- Metadata config: `src/config/metadata.ts`
- Static fallback image: `public/preview.webp`

## Estimate

4 days total. 3d for the OG image endpoint + post wiring, 0.5d for profile metadata, 0.5d for cross-platform testing.

## Assignee

@aldertnl  (design), @Nirzhuk (OG image endpoint + wiring), @secondl1ght (profile metadata + platform QA)

### LINKED ISSUE #2168: [Bug]: Can open Collection in post page by editing url
### Description

Navigating to a collection using the post url opens the collection in post page. There's no use for this in the app because it cannot be navigated to via the UI. (Although it does allow you to have a secret reply thread)

### Expected Behaviour

Redirect from `/post/*` to `/collection/*` when kind is collection.

### Steps to Reproduce

1. Find a collection
2. Change it's url from `/collection/*` to `/post/*`
3. Observe the collection loads in post page

### Screenshots

https://github.com/user-attachments/assets/78521fd7-cbf1-45a9-bb8c-87fa4684cdd2

### Console Log

```shell

```

### Browser & Version

_No response_

### PR DESCRIPTION
**Pull request by secondl1ght**
## ✨ Dynamic Open Graph / Twitter preview images

  Adds per-URL dynamic social preview cards for the **post**, **profile**, and **collection** routes using Next.js's
  native `opengraph-image` / `twitter-image` file convention with `ImageResponse` from `next/og`. Previously every shared
  link rendered a single static `/preview.webp`; now shared links show the actual author, content, and avatar.

  ### 🖼️  What renders

  **Post** (`/post/[userId]/[postId]`)
  - **Text** variant — avatar + name, centered post text, `pubky.app` footer
  - **Image** variant (`kind: image`) — caption + the post's image (resolved from the attachment)
  - **Article** variant (`kind: long`) — 📰 newspaper icon + article title, with a plain-text excerpt of the (Markdown)
  body

  **Profile** (`/profile/[pubky]`)
  - Large avatar + name + bio + follower / post counts
  - Now also emits **dynamic title/description + canonical** (was fully static before)

  **Collection** (`/collections/[userId]/[postId]`)
  - ⏳ Deferred (design pending): all wiring is in place, but `renderCollectionOg` currently serves the default static
  preview. Re-activating is a single-function change in a follow-up.

  ### 🧠 Behavior & decisions

  - **Server-safe rendering** — reuses/extends the existing Nexus fetch path (no Dexie/IndexedDB on the server) and only
  pure URL builders, to keep client-side code out of the server bundle.
  - **Caching** — segment `revalidate = 3600` aligned with the upstream fetch, plus a `stale-while-revalidate`
  cache-control header so CDNs/crawlers cache aggressively.
  - **Graceful fallbacks** — every remote image is pre-fetched in catchable code (so a broken image never 500s a crawler);
  on missing/errored data the route redirects to the app's configured default preview image. Missing avatar → the
  brand-colored Facehash "no image" face (inlined as SVG, crisp at 80px and 384px).
  - **Display name** — nameless users fall back to a shortened public key (reuses the app's existing `formatPublicKey`).
  - **Content-less posts** (e.g. simple reposts) — still surface the `{name} on Pubky` title + dynamic image; the
  description is suppressed (`null`) rather than inheriting the root layout's generic tagline, so it doesn't read as if
  the author wrote it.
  - **Deleted posts** — short-circuited before content parsing, so they render the "deleted" notice deterministically.
  - **Markdown** — article body is stripped to clean plain text via a new dependency-free `markdownToText()` helper
  (headings, emphasis, links/images, code, blockquotes, lists, tables, HTML, escaped punctuation; never throws).
  - **Text overflow** — `word-break: break-word` on all wrapping text, plus explicit `maxHeight` line caps (satori's
  `-webkit-line-clamp` is a no-op in this `@vercel/og` build) so long/URL-heavy content stays within the card: post text 3
  lines, article body 2, profile bio 3, image caption 2.
  - **Collection redirect** — collection-kind posts on `/post` redirect to `/collections` + set `alternates.canonical`.
  It's a client-side (soft) redirect: the app's cookie-based root layout forces dynamic streaming app-wide, so a hard 308
  isn't achievable without middleware (intentionally deferred). Crawlers still get the correct preview since the `/post`
  image route delegates collection kind to the collection renderer.

  ### 🔒 Security (SSRF)

  `resolvePostAttachmentUrl` accepts **only `pubky://` file URIs** (which resolve to our CDN) and rejects everything else,
  including absolute `http(s)` URLs — so the server-side image fetch can never be pointed at an arbitrary/internal host.
  Every genuine attachment is a homeserver file URI, so this reflects reality with **zero impact on legitimate image
  rendering** (verified: real image posts still render their image). The text side has no injection surface — output is a
  rasterized PNG, and text is escaped/treated as plain text nodes.

  ### 🔧 Supporting changes

  - **`Metadata()` helper** — added `omitImages` (lets the file-convention image be the single image source, avoiding a
  static-vs-dynamic conflict) and made `description` optional (empty → `null`/`''` so an empty page doesn't inherit the
  parent's generic description).
  - **Shared OG constants** — the route files consume `OG_SIZE` / `OG_CONTENT_TYPE` (`revalidate` stays a literal, since
  Next requires segment config to be statically analyzable).
  - **Image transcoding** — fetched CDN images (WebP) are transcoded to PNG with `sharp` (satori can't decode WebP) and
  downscaled to bound the embedded payload; uses the `feed` file variant since the image only ever renders in a small
  preview card.
  - **`sharp`** — promoted from a transitive dependency (already installed via Next's image optimization) to an explicit
  direct dependency; the app's existing image utilities are browser/Canvas-only and can't run server-side.
  - **Fonts** — bundled Inter Tight Medium/Bold TTFs, lazily loaded once per process.
  - Extracted reusable helpers (`truncateByGraphemes`, `deriveTextPreview`) and exported `fetchWithValidation` for reuse
  across the metadata + OG data layers.

  ### ✅ Testing

  - Unit tests for the pure/logic pieces: grapheme truncation, post-preview derivation, Markdown stripping (edge cases),
  attachment/avatar URL resolution (incl. the SSRF rejection guard), profile fetch (counts-404 → zeros), image transcode
  guards, `Metadata` (`omitImages` + optional description), the fallback redirect, structural snapshots of the OG
  components, and each route's `generateMetadata` + the collection redirect.
  - Verified end-to-end against real Nexus staging data and visually QA'd against the Figma designs across every variant
  (text, image, article, profile with/without avatar/bio, fallbacks, reposts, deleted posts, long-URL overflow).

  _This pull request summary was generated, for full implementation details please reference the file changes._

### CODE REVIEW

File: src/libs/og/ogData.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    const res = await fetch(url, { next: { revalidate: OG_REVALIDATE } });
+    if (!res.ok) return null;
+    const contentType = res.headers.get('content-type') ?? '';
+    if (!contentType.startsWith('image/')) return null;
+
+    const input = Buffer.from(await res.arrayBuffer());
+    const png = await sharp(input)
+      .resize({ width: OG_IMAGE_MAX_EDGE, height: OG_IMAGE_MAX_EDGE, fit: 'inside', withoutEnlargement: true })
+      .png()
+      .toBuffer();
```
[the remark this answers is not in the document]
Comment (secondl1ght) [reply src/libs/og/ogData.ts:135]: Our CDN already handles image optimization on ingest and it creates asset variants. In addition we use the `feed` variant for these OG images which is the smaller version.

**Comment by secondl1ght** [comment]
@greptile responded to your comment

**Comment by secondl1ght** [comment]
@greptile so how do we update your confidence score? 😄 Only re-view does that?

**Comment by secondl1ght** [comment]
@greptile no code has changed but your comment has been addressed. The review summary still shows outdated information.

**Review by infin1t3 — CHANGES_REQUESTED** [review]
I walked through the preview flow on the ephemeral deployment and found two metadata issues and one potential image rendering issue. I’ve attached a recording showing the complete reproduction.

https://drive.google.com/file/d/1EaydpFTx15bqhVBD7gmvMckUv83DwNKB/view?usp=sharing

## 1. Metadata stays stale after client-side navigation

### Reproduction

1. Open the home page.
2. Navigate to a post without refreshing the page.
3. Run the following in DevTools:

```js
const read = (selector, attribute) =>
  document.querySelector(selector)?.getAttribute(attribute);

console.table({
  canonical: read('link[rel="canonical"]', 'href'),
  ogUrl: read('meta[property="og:url"]', 'content'),
  ogTitle: read('meta[property="og:title"]', 'content'),
  ogDescription: read(
    'meta[property="og:description"]',
    'content',
  ),
  ogImage: read('meta[property="og:image"]', 'content'),
  twitterImage: read(
    'meta[name="twitter:image"]',
    'content',
  ),
});
```

After navigating from **Home** to a post, the document still contains the homepage metadata:

```text
og:title      = Pubky App – Unlock the web
og:image      = https://pubky.app/preview.webp
twitter:image = https://pubky.app/preview.webp
og:url        = https://pubky.app
canonical     = https://pubky.app
```

After refreshing the same post URL and running the command again, the post-specific title, description, `opengraph-image`, and `twitter-image` appear correctly.

It looks like client-side navigation through the intercepted post route does not run (or does not receive) the metadata from the full `/post/[userId]/[postId]` route, leaving the previous page metadata in the document until a hard reload.

Social crawlers normally perform a full document request, so this may not break external unfurling, but it does leave the browser document metadata incorrect after normal in-app navigation.

---

## 2. `canonical` and `og:url` are incorrect even after refresh

After a hard refresh, the dynamic title, description, and images are correct, but these values still remain:

```text
canonical = https://pubky.app
og:url    = https://pubky.app
```

They should instead identify the specific post, for example:

```text
https://pubky.app/post/<userId>/<postId>
```

This directly affects canonical page identity and Open Graph object identity. Different posts currently identify themselves as the Pubky homepage.

---

The image renderers for regular posts themselves look good: both `/opengraph-image` and `/twitter-image` return the correct 1200×630 post/article cards.

For image posts however, image rendered looks a bit cut-off from both top and the bottom - you can also see that in the video.

**Comment by secondl1ght** [comment]
@infin1t3 thanks for the review comments, for 1 that is not a behavior this PR introduced and you're right that it is because of the intercepted route. Shouldn't be an issue though because like you said the crawlers never hit that path. If we want the title to update on client-side navigation we could create a low prio bug report for that one.

For 2, this is another pre-existing one and I have fixed it in the latest commit. About the image question, what we have here matches the design. I was told the reason to crop out part of the image was because it causes "curiosity" and I think they used FB as an example (discussion was on Slack).

**Review by infin1t3 — APPROVED** [review]
